analyzers/ImpactParameter.py

```diff
@@ -1,48 +1,214 @@
         assumed_vertex = TVector3(0, 0, 0)
         jets = getattr(event, self.cfg_ana.jets)
         for jet in jets:
-            #for ptc in jet.constituents[211]:
-            #    IPsn = 1
-            #    N_charge = 0
-                #for i in range(len(const)):
-                #    if const[i][1][0]._charge <> 0:
-                #        if len const[i][1] > 0:
-                #            for l in range(len(const[i][1]))
-                #                partic = const[i][1][l]
-                #                h = partic.path
-                #                Helix.compute_IP(h, assumed_vertex)
-                #                setattr(partic, self.IP, h.IP)
-                #                setattr(partic, self.tIP, h.closest.t)
-                #                setattr(partic, self.xIP, h.closest.x)
-                #                setattr(partic, self.yIP, h.closest.y)
-                #                setattr(partic, self.zIP, h.closest.z)
-                #                IPsn *= h.IP
-                #                N_charge += 1
-                #        else :
-                #            pass
-                #    else :
-                #        pass
-                #if N_charge <> 0 :
-                #    IPs = IPSn**(1.0/N_charge)
-                #else :
-                #    pass
+            IP_b_LL = 0     # value of the log likelihood ratio based on IP initiated at 0
+            IPs_b_LL = 0    # value of the log likelihood ratio based on IP significance initiated at 0
+            ipsig_ptcs = [] # list of IP signif and associated ptcs
             for id, ptcs in jet.constituents.iteritems():
-                if id in [22,130]:
+                if abs(id) in [22,130,11]:
                     continue
                 for ptc in ptcs :
-                    if ptc._charge == 0 :
-                        continue
-                    ptc.path.compute_IP(assumed_vertex)
-                    #jet.IP = ptc.path.IP
-            #setattr(event, self.cfg_ana.output, IPs)
+                    if hasattr(ptc.path, "charge") :
+                        if ptc.path.charge == 0 :
+                            continue
+                        ptc.path.compute_IP(assumed_vertex,jet)
+                        
+                        ptc_IP_signif = 0
+                        if hasattr(ptc.path, 'points') == True and 'beampipe_in' in ptc.path.points:
+                            ptc.path.compute_theta_0()
+                            ptc.path.compute_IP_signif(ptc.path.IP,
+                                                       ptc.path.theta_0,
+                                                       ptc.path.points['beampipe_in'])
+                        else :
+                            ptc.path.compute_IP_signif(ptc.path.IP, None, None)
+                        
+                        if ptc.path.p4.Perp() > 1 and (ptc.path.IPx**2 + ptc.path.IPy**2)**0.5 < 0.002 and ptc.path.IPz**2 < 0.17**2 :
```

Comment:
avoid hardcoded numbers. 
Instead use parameters in the cfg_ana: 
cfg_ana.dxy_max 
cfg_ana.dz_max